_posts/2018-09-12-graphql-is-not-odata.md

```diff
@@ -0,0 +1,224 @@
+
+The conversation took off from there. We talked about the flaws of OData, the positives of GraphQL, and the downsides of GraphQL too. We didn't explicitly cover the positives of OData, but some of them were covered in side threads. We'll review all of these angles here.
+
+### Flaws in OData
+
+*[see in thread](https://twitter.com/JeffHandley/status/1034829487794352128)*
+
+In summary, OData was a way to serialize a SQL statement into a URL. When first applied, it gave too much power to the client, allowing it to use joins, where clauses, and sorting. A lot of that power can be dialed back now, but I still found it challenging to limit the exposure.
+
+By opening up an OData endpoint, you lose control of what your clients are able to do and you don't have explicit queries or use cases tracked. So when you need to change your schema, it becomes challenging to maintain backward compatibility with what your clients are doing.
+
+It also becomes exceptionally difficult to optimize performance for any given use case because all execution runs through the same pipeline. Hooks were there to do it, but you have to process IQueryables in order to fulfill specific queries with tuned implementations. So when a client conducts a specific query and reports to you that performance is poor, you have to do really heavy lifting to address it. And in doing so, you can inadvertently disrupt other callers with similar but inexact scenarios/queries.
+
+In many ways, OData queries result in remote code execution by design. My team and I had to address this specific security bulletin where specially-crafted GET requests could employ remote code execution to take down any public OData server: [https://docs.microsoft.com/en-us/security-updates/SecurityBulletins/2013/ms13-007](https://docs.microsoft.com/en-us/security-updates/SecurityBulletins/2013/ms13-007).
+
+### Positives about GraphQL
+
+*[see in thread](https://twitter.com/JeffHandley/status/1035075407043690496)*
+
+The server implements resolvers that fulfill specific graph queries—the client cannot ask for anything the server does not explicitly handle. The server returns only what the client asked for. No more, no less. No wasted bits over the wire.
+
+APIs sitting behind GraphQL do not need to worry about serving every possible shape clients will want. Web wants more, mobile wants less? APIs don’t care. Clients ask for what they need. APIs return domain models. GraphQL transforms into View Models.
+
+The GraphQL layer can then specifically optimize queries over the RESTful APIs, minimizing API calls necessary to resolve a query. Yes, it is work to implement each resolver, but the explicitness is so liberating. And your clients have no idea what is happening behind the GraphQL façade. One call from the client could fan out to a dozen APIs (in a good way*).  Orchestration stays out of the UI!
```

Comment:
on the notion of liberating, would add in the same way that REST tamed the openness of the Web through the formalization of constraints, GraphQL cleared up many ambiguities in REST by further constraining the uniform interface constraint.

Reply:
Going to add this as a quote from you